### PR TITLE
apply changes for moe pic

### PR DESCRIPTION
**Pull request by rebel-wonsubkim**
### 🚀 Summary of Changes

> *What does this PR do? What feature, fix, or improvement does it bring?*

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`bug-fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

File: examples/test.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Copyright 2025 Rebellions Inc. All rights reserved.
```
Comment (rebel-jiwoopark) [inline examples/test.py:1]: `test.py`
It would be better to give the test script a more descriptive filename so that its role can be understood more easily.

Comment (rebel-wonsubkim) [reply examples/test.py:1]: remove this file :)

File: vllm_rbln/attention/backends/flash_attention.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# ruff: noqa
```
Comment (rebel-jiwoopark) [inline vllm_rbln/attention/backends/flash_attention.py:15]: Could you explain why the linting is being ignored?

Comment (rebel-wonsubkim) [reply vllm_rbln/attention/backends/flash_attention.py:15]: okay I will restore linting and modify codes to meet constraint

File: examples/test.py (hunk 1 above)
Comment (rebel-jiwoopark) [reply examples/test.py]: thank you!

**Comment by rebel-wonsubkim** [comment]
@rebel-jiwoopark okay

**Comment by rebel-jiwoopark** [comment]
I’d appreciate it if you could merge promptly once the conflict is resolved.

File: vllm_rbln/worker/model_runner.py [hunk 3]
```diff
@@ -471,10 +491,43 @@ def decoder_layer_block(self, decoder_layer, hidden_states, residual,
...
+                inputs_embeds=inputs_embeds)
+
+            if get_pp_group().is_last_rank:
+                # last rank create real model output
+                if selected_token_indices is not None:
+                    # aten::select -> adv_index -->
+                    #     contrib_dynamic_take (tensor -> scalar)
+                    # aten::index_select --> take -->
+                    #     contrib_dynamic_take (tensor -> scalar)
+                    model_output = model_output[:, selected_token_indices]
```
Comment (huijjj) [inline vllm_rbln/worker/model_runner.py:523]: Can rebel compiler handle this kind of dynamicity in selected_token_indices? or will it just recompile every time when it encounters unseen selected_token_indices?

Comment (huijjj) [reply vllm_rbln/worker/model_runner.py:523]: This tensor changing dynamically would also affect the subsequent matrix multiplication (lm_head). But since it’s the first dimension, which can be seen as a kind of batch dimension in matrix multiplication, you mean this falls under one of the patterns that can be handled. That’s interesting, thank you for clarifying!

**Review by huijjj — COMMENTED** [review]
Recent changes addressing issues around block allocation (https://github.com/rebellions-sw/vllm-rbln/pull/77, https://github.com/rebellions-sw/vllm-rbln/pull/82) seems to be omitted during the rebase, please review and include these changes to ensure proper functionality.

**Review by rebel-jiwoopark — CHANGES_REQUESTED** [review]
please check pr #77 #82.

File: vllm_rbln/worker/model_runner.py (hunk 3 above)
Comment (huijjj) [reply vllm_rbln/worker/model_runner.py:523]: Yes, selected_token_indices remains single element tensor as you mentioned. 

But I was questioning about the sliced tensor `model_output[:, selected_token_indices]` whose shape changes accordingly to the value of `selected_token_indices` leading to dynamicity in following matrix multiplication conducted in `compute_logits`.

I have created an simple test script to check this out by myself but my code somewhat seems to fallback to CPU.
```
import torch
import rebel 

def foo(input, slice_index, weight):
    return input[:slice_index]@weight

foo = torch.compile(foo, backend="rbln", dynamic=False)

hidden_state = torch.randn(64, 2048)
weight = torch.randn(2048, 4096)

for i in range(1, 63):
    _ = foo(hidden_state, torch.tensor([i], dtype=torch.int32), weight)

assert len(torch._dynamo.utils.get_compilation_metrics()) == 1 # rbln torch.compile backend gets triggered but somewhat never compiles the graph for device acceleration
```
Please correct me if I am missing something.

**Comment by rebel-wonsubkim** [comment]
@huijjj @rebel-jiwoopark we have already considered a dummy block when calculating max_required_num_blocks
it is related to #77

File: vllm_rbln/worker/worker.py [hunk 4]
```diff
@@ -286,37 +299,43 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
...
+                # quantization : 4 (This is an ad-hoc value. Need to fix it)
+                nbits_per_param=16
+                if not self.model_config.quantization else 4,
+                n_model_params=sum(
+                    p.numel() for p in self.model_runner.model.parameters()),
+                # 2 : 1 for prefill and decode each
+                num_runtimes=2) - 1
+
+            max_required_num_blocks = (num_blocks_per_seq *
+                                       self.scheduler_config.max_num_seqs) + 1
```
Comment (huijjj) [inline vllm_rbln/worker/worker.py:330]: You should account for [V0 kv cache manager heuristics](https://github.com/vllm-project/vllm/blob/v0.9.1/vllm/core/block_manager.py#L207-L234) here. Please refer to this [commit](https://github.com/rebellions-sw/vllm-rbln/pull/77/commits/923093d3408ed1040549f6d83979f590aa34a0a1) for detail, setting block numbers like this yields prefill-only text generation.

File: vllm_rbln/worker/model_runner.py (hunk 3 above)
Comment (huijjj) [reply vllm_rbln/worker/model_runner.py:523]: Sorry, I made a confusion. Tensor shape does not change as `model_output[:, selected_token_indices]` is an index_select not slice.

Comment (rebel-wonsubkim) [reply vllm_rbln/worker/model_runner.py:523]: @huijjj input[:slice_index] is actual dynamic slicing
test code - model_output[: selected_token_indices] --> actual dynamic slicing, it is not captured by torch dynamo
our code - model_output[:, selected_token_indices] --> it can be captured since it is not dynamic slicing

File: vllm_rbln/worker/worker.py [hunk 5]
```diff
@@ -286,37 +299,43 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
...
+            max_required_num_blocks = max_required_num_blocks * ve_cnt
+
+            num_gpu_blocks = min(max_num_blocks, max_required_num_blocks)
```
[the remark this answers is not in the document]
Comment (huijjj) [reply vllm_rbln/worker/worker.py:333]: I've missed the details in attention backend. _allocate_kv_cache allocates 1 extra block for padding as get_kv_cache_shape returns the shape with 1 extra block, Sorry again for the confusion.

**Review by huijjj — APPROVED** [review]
This changes correctly addresses the issue mentioned in previous PRs (https://github.com/rebellions-sw/vllm-rbln/pull/77, https://github.com/rebellions-sw/vllm-rbln/pull/82), LGTM

@rebel-jiwoopark PTAL

**Review by rebel-jiwoopark — APPROVED** [review]
